src/interp/interp-wasi.cc

```diff
@@ -76,27 +127,112 @@ class WasiInstance {
     return Result::Ok;
   }
 
+  Result environ_get(const Values& params, Values& results, Trap::Ptr* trap) {
+    uvwasi_size_t environc;
+    uvwasi_size_t environ_buf_size;
+    uvwasi_environ_sizes_get(uvwasi, &environc, &environ_buf_size);
+    uint32_t wasm_buf = params[1].i32_;
+    char* buf = getMemPtr<char>(wasm_buf, environ_buf_size, trap);
+    if (!buf) {
+      return Result::Error;
+    }
+    char** host_env = new char*[environc];
+    uvwasi_environ_get(uvwasi, host_env, buf);
+
+    // Copy host_env pointer array wasm_env)
+    for (uvwasi_size_t i = 0; i < environc; i++) {
+      uint32_t rel_address = host_env[i] - buf;
+      uint32_t dest = params[0].i32_ + (i * sizeof(uint32_t));
+      CHECK_RESULT(writeValue<uint32_t>(wasm_buf + rel_address, dest, trap));
+    }
+
+    delete[] host_env;
+    return Result::Ok;
+  }
+
+  Result environ_sizes_get(const Values& params,
+                           Values& results,
+                           Trap::Ptr* trap) {
+    uvwasi_size_t environc;
+    uvwasi_size_t environ_buf_size;
+    uvwasi_environ_sizes_get(uvwasi, &environc, &environ_buf_size);
+    CHECK_RESULT(writeValue<uint32_t>(environc, params[0].i32_, trap));
+    CHECK_RESULT(writeValue<uint32_t>(environ_buf_size, params[1].i32_, trap));
+    if (trace_stream) {
+      trace_stream->Writef("environ_sizes_get -> %d %d\n", environc, environ_buf_size);
+    }
+    return Result::Ok;
+  }
+
+  Result args_get(const Values& params, Values& results, Trap::Ptr* trap) {
+    uvwasi_size_t argc;
+    uvwasi_size_t arg_buf_size;
+    uvwasi_args_sizes_get(uvwasi, &argc, &arg_buf_size);
+    uint32_t wasm_buf = params[1].i32_;
+    char* buf = getMemPtr<char>(wasm_buf, arg_buf_size, trap);
+    if (!buf) {
+      return Result::Error;
+    }
+    char** host_args = new char*[argc];
+    uvwasi_args_get(uvwasi, host_args, buf);
+
+    // Copy host_args pointer array wasm_args)
+    for (uvwasi_size_t i = 0; i < argc; i++) {
+      uint32_t rel_address = host_args[i] - buf;
+      uint32_t dest = params[0].i32_ + (i * sizeof(uint32_t));
+      CHECK_RESULT(writeValue<uint32_t>(wasm_buf + rel_address, dest, trap));
+    }
+
+    delete[] host_args;
+    return Result::Ok;
+  }
+
+  Result args_sizes_get(const Values& params,
+                        Values& results,
+                        Trap::Ptr* trap) {
+    uvwasi_size_t argc;
+    uvwasi_size_t arg_buf_size;
+    uvwasi_args_sizes_get(uvwasi, &argc, &arg_buf_size);
+    CHECK_RESULT(writeValue<uint32_t>(argc, params[0].i32_, trap));
+    CHECK_RESULT(writeValue<uint32_t>(arg_buf_size, params[1].i32_, trap));
+    if (trace_stream) {
+      trace_stream->Writef("args_sizes_get -> %d %d\n", argc, arg_buf_size);
+    }
+    return Result::Ok;
+  }
+
   // The trace stream accosiated with the instance.
   Stream* trace_stream;
 
   Instance::Ptr instance;
+
  private:
+  template <typename T>
+  Result writeValue(T value, uint32_t target_address, Trap::Ptr* trap) {
```

Comment:
nit: style is `WriteValue` (should probably fix `GetMemPtr` too). The WASI names are fine to keep as is.

Reply:
Yeah, your right.   I'm was just being laz.....  i was just doing TTD and test only use one :)